### DISCUSSION TITLE
What's the new guideline for using metrics?

### DISCUSSION THREAD

**Discussion by christiannagel** [microsoft/aspire#3313] [Q&A] [answered]
Because the OpenTelemetry metrics library is not stable yet (e.g. https://github.com/dotnet/aspire/pull/3087), the configuration from the ServiceDefaults project, and from the components has been removed. What is the new guideline to configure what metrics should be shown with the dashboard, also custom meters?

**Comment by davidfowl** [comment]
Specifically for event counters right?

**Reply by christiannagel** [reply]
Yes, specifically for event counters. The Aspire components no longer have the `Metrics` property. To turn metrics on with these components we now must search for the activity source names and turn it on explicitly - is this correct?

With custom metrics values I've now seen that no changes are needed with the code to see them in the dashboard.

**Reply by danmoseley** [reply]
Cc @JamesNK

**Reply by davidfowl** [reply]
Aspire component still have metrics, just not event counter based metrics. The native metric APIs (Meter) are still turned on by default if a component provides them.

**Reply by christiannagel** [reply]
Now I see, EF Core is still using the old EventSource API (https://github.com/dotnet/efcore/blob/45448efb1da8914489739bc4116f7a8f6c9374a2/src/EFCore/Infrastructure/EntityFrameworkEventSource.cs#L11) and not the new Meter API. 
Metrics from components using the EventSource API, such as the Microsoft Entity Framework Core Cosmos DB component is no longer available, that's why the `Metrics` property has been removed with these components (https://github.com/dotnet/aspire/commit/364ce504155c4bcc9824309fcda067dd8438c073) - is this correct?

I've been using the `Metrics` property with these components, and was thinking it's been removed with all components, but this seems not to be the case.

What's the reason for this change?
Going forward, all the components (e.g. EF Core) using event counter-based metrics (EventSource) should switch to the new Meter API, is this correct?

**Reply by JamesNK** [reply, marked as answer]
`OpenTelemetry.Instrumentation.EventCounters` is a pre-release package. We can't keep referencing once Aspire components are not pre-release.

I think eventually EF Core and friends will publish real metrics, but until then you'll have to reference the event counters package and add the counters in your app.

**Reply by christiannagel** [reply]
Thanks @JamesNK! I wasn't aware OpenTelemetry.Instrumentation.EventCounters is the only option to use event counters. `dotnet counters` supports monitoring with EventSource counters as well. Before .NET 6, event counters were the only option we had with .NET.

More reasons to switch to the new Meter API :-)

With EF Core I've seen it's in the .NET 9 milestone: https://github.com/dotnet/efcore/issues/25880
